rust/operator-binary/src/controller.rs

```diff
@@ -0,0 +1,490 @@
+            .build(),
+        spec: Some(ServiceSpec {
+            ports: Some(vec![ServicePort {
+                name: Some("hive".to_string()),
+                port: APP_PORT.into(),
+                protocol: Some("TCP".to_string()),
+                ..ServicePort::default()
+            }]),
+            selector: Some(role_selector_labels(hive, APP_NAME, &role_name)),
+            type_: Some("NodePort".to_string()),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+/// The rolegroup [`ConfigMap`] configures the rolegroup based on the configuration given by the administrator
+fn build_metastore_rolegroup_config_map(
+    rolegroup: &RoleGroupRef<HiveCluster>,
+    hive: &HiveCluster,
+    metastore_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<ConfigMap> {
+    let mut hive_site_data = String::new();
+    let log4j_data = include_str!("../../../deploy/external/log4j.properties");
+
+    for (property_name_kind, config) in metastore_config {
+        match property_name_kind {
+            PropertyNameKind::File(file_name) if file_name == HIVE_SITE_XML => {
+                let mut data = BTreeMap::new();
+
+                for (property_name, property_value) in config {
+                    data.insert(property_name.to_string(), Some(property_value.to_string()));
+                }
+
+                // TODO: make configurable
+                data.insert(
+                    "hive.metastore.warehouse.dir".to_string(),
+                    Some("/stackable/warehouse".to_string()),
+                );
+
+                hive_site_data =
+                    stackable_operator::product_config::writer::to_hadoop_xml(data.iter());
+            }
+            _ => {}
+        }
+    }
+
+    ConfigMapBuilder::new()
+        .metadata(
+            ObjectMetaBuilder::new()
+                .name_and_namespace(hive)
+                .name(rolegroup.object_name())
+                .ownerreference_from_resource(hive, None, Some(true))
+                .context(ObjectMissingMetadataForOwnerRef)?
+                .with_recommended_labels(
+                    hive,
+                    APP_NAME,
+                    hive_version(hive)?,
+                    &rolegroup.role,
+                    &rolegroup.role_group,
+                )
+                .build(),
+        )
+        .add_data(HIVE_SITE_XML, hive_site_data)
+        .add_data(LOG_4J_PROPERTIES, log4j_data)
+        .build()
+        .with_context(|| BuildRoleGroupConfig {
+            rolegroup: rolegroup.clone(),
+        })
+}
+
+/// The rolegroup [`Service`] is a headless service that allows direct access to the instances of a certain rolegroup
+///
+/// This is mostly useful for internal communication between peers, or for clients that perform client-side load balancing.
+fn build_metastore_rolegroup_service(
+    rolegroup: &RoleGroupRef<HiveCluster>,
+    hive: &HiveCluster,
```

Comment:
For consistency reasons I would switch the order of these two arguments because the &HiveCluster is passed as first (and sometimes only) argument everywhere else.

Reply:
yeah done